### PR TITLE
[ADHOC] feat(cli): `seed` command for creating boosts

### PR DESCRIPTION
**Pull request by topocount**
still needs:

- [x] implement onboarding that generates default config - @topocount 
- [x] implement `--from` flag that pulls the seed config - @topocount 
- [x] route actions to the correct constructor - @sammccord working on typing for this
- [x] fix signatures to parse out hex from abi items - @topocount

resolves BOOST-4941
resolves BOOST-4940

### CODE REVIEW

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 82497609b8b9c711383b155171600eca59705111
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 38ffe444e080dffb4bb1036de590bd1dc53e93c1
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9c30f84a6b7dde06b54ec3324303b31c2bf769ce
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f79b385f590cd2afeddf8706c9990be6a771e204
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against abf0255e0bc32dcaad43919f9cf9225282bea289
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 396b1fb263c9d095469496323af3d8901dba2ce8
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7d113d4bbfb4cbe2a41f34016825dd2cb398c1e3
</p>

File: packages/cli/src/help.ts [hunk 1]
```diff
@@ -4,14 +4,17 @@ Deploy Boost Protocol V2, and retrieve base implementation addresses.
...
+  seed first.json [...]
 
 Options:
   -h, --help     Print help
   -v, --version  Print version
 
   --chain        Viem chain name, see https://github.com/wevm/viem/tree/main/src/chains/definitions for a list of valid names
-  --privateKey   Private key to use for contract operations, see https://viem.sh/docs/accounts/local/privateKeyToAccount#generating-private-keys for more information
+  --privateKey   Private key to derive account from to use for contract operations, see https://viem.sh/docs/accounts/local/privateKeyToAccount#generating-private-keys for more information
+  --mnemonic     Mnemonic to derive account from to use for contract operations, see https://viem.sh/docs/accounts/local/mnemonicToAccount#generating-private-keys for more information
```
Comment (sammccord) [inline packages/cli/src/help.ts:17]:
```suggestion
  --mnemonic     Mnemonic to derive account from to use for contract operations, see https://viem.sh/docs/accounts/local/mnemonicToAccount#generating-private-keys for more information
  --rpcUrl       Optional URL to provide to Viem http transport
```

File: packages/cli/package.json [hunk 2]
```diff
@@ -15,42 +15,44 @@
...
     "dev": "vite build --watch",
     "clean": "rm -rf dist"
   },
   "optionalDependencies": {
-    "@boostxyz/sdk": "workspace:*",
-    "@boostxyz/evm": "workspace:*"
+    "@boostxyz/sdk": "*"
+  },
+  "dependencies": {
+    "@inquirer/prompts": "^7.0.1"
```
Comment (sammccord) [inline packages/cli/package.json:56]:
```suggestion
```

File: packages/cli/package.json [hunk 3]
```diff
@@ -15,42 +15,44 @@
...
-    "build": "vite build && tsc --build --emitDeclarationOnly --declaration --declarationMap --force",
+    "build": "vite build",
+    "check:ts": "tsc --build --emitDeclarationOnly --declaration --declarationMap --force",
     "dev": "vite build --watch",
     "clean": "rm -rf dist"
   },
   "optionalDependencies": {
-    "@boostxyz/sdk": "workspace:*",
-    "@boostxyz/evm": "workspace:*"
+    "@boostxyz/sdk": "*"
```
Comment (sammccord) [inline packages/cli/package.json:53]:
```suggestion
    "@boostxyz/sdk": "workspace:*"
```

File: packages/cli/src/commands/seed.ts [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+  });
+
+  const chainId = chain.id;
+
+  if (positionals.at(0) === 'generate') {
+    return makeSeed({ account: account?.address, chainId });
+  }
+
+  if (positionals.at(0) === 'erc20') {
+    let erc20 = new MockERC20(defaultOptions, {});
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:88]:
```suggestion
    let erc20 = new MockERC20({ config, account }, {});
```

File: packages/cli/src/help.ts [hunk 5]
```diff
@@ -4,14 +4,17 @@ Deploy Boost Protocol V2, and retrieve base implementation addresses.
 Usage: boost <COMMAND>
 
 Commands:
-  deploy               Deploy a new version of the protocol
+  deploy                  Deploy a new version of the protocol
+  seed generate           Generate a new basic Boost seed to stdout
```
Comment (sammccord) [inline packages/cli/src/help.ts:8]:
```suggestion
  seed generate           Generate a new basic Boost seed
  seed erc20              Deploys a new ERC20
```

File: packages/cli/src/commands/seed.ts [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+
+  let sharedBudget: ManagedBudget | undefined;
+  let sharedBudgetConfig: ManagedBudgetPayload | undefined;
+  const boostIds: string[] = [];
+  for (const template of templates) {
+    let budget: ManagedBudget;
+    if (typeof template.budget === 'string' && isAddress(template.budget))
+      budget = core.ManagedBudget(template.budget);
+    // TODO: create budget from Core
+    else if (sharedBudgetConfig === template.budget && sharedBudget) {
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:132]: I'm not sure `sharedBudgetConfig === template.budget` will ever be true, the template.budget will be a different object between each template

File: packages/cli/src/commands/seed.ts [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+    const boost = await core.createBoost({
+      protocolFee: template.protocolFee,
+      maxParticipants: template.maxParticipants,
+      budget: budget,
+      action: core.EventAction(template.action as EventActionPayload),
+      validator: core.SignerValidator(template.validator),
+      allowList: await getAllowList(template, { core }),
+      incentives,
+    });
+    boostIds.push(boost.id.toString());
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:187]: boost ids are probably fine for now, but this makes me want to write a `boost.serialize` that returns a boost to a shape like boostconfig

File: packages/cli/src/help.ts [hunk 8]
```diff
@@ -4,14 +4,17 @@ Deploy Boost Protocol V2, and retrieve base implementation addresses.
...
+  seed first.json [...]
```
Comment (sammccord) [inline packages/cli/src/help.ts:9]:
```suggestion
  seed seed.json [...]    Create Boosts using JSON configuration files generable by `seed generate`. 
                          You can pass as many file paths as needed, and each will be validated before any operations occur.
                          Budgets will be automatically funded based on the incentives configuration, using accounts derived from the configured private key or mnemonic.
```

File: packages/cli/src/cli.ts [hunk 9]
```diff
@@ -97,15 +100,15 @@ async function main() {
...
-  if (!cachedResult) {
+  if (typeof cachedResult == undefined) {
     throw new Error(
       `No result found after cache and command operations. This shouldn't happen.`,
     );
   }
 
   // Cache result
-  await fs.mkdir(cacheDir, { recursive: true });
-  await fs.writeFile(cachedResultPath, cachedResult, { encoding: 'utf8' });
```
Comment (sammccord) [inline packages/cli/src/cli.ts:108]: yeah we don't reallllly need caching....

File: packages/cli/src/commands/seed.ts [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+  }
+
+  return {
+    boostIds,
+  };
+};
+
+async function fundBudgetForIncentive(
+  budget: Budget,
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:196]:
```suggestion
  options: DeployableOptions,
  budget: Budget,
```

File: packages/cli/src/commands/seed.ts [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+  amount: bigint,
+  asset: Address,
+) {
+  if (asset && amount) {
+    let erc20 = new MockERC20(defaultOptions, asset);
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:201]:
```suggestion
    let erc20 = new MockERC20(options, asset);
```

File: packages/cli/src/commands/seed.ts [hunk 12]
```diff
@@ -0,0 +1,504 @@
...
+    await fundBudget(
+      defaultOptions,
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:203]:
```suggestion
      options,
```

File: packages/cli/src/commands/seed.ts [hunk 13]
```diff
@@ -0,0 +1,504 @@
...
+}
+
+async function fundBudget(
+  options: DeployableTestOptions,
+  erc20: MockERC20,
+  budget: Budget,
+  amount = parseEther('110'),
+) {
+  //await fundErc20(options, erc20, [], parseEther(amount.toString()))();
+  console.log(`minting ${amount} to ${options.account.address}`);
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:426]:
```suggestion
```
this'll mess with the stdout and end up looking lik
```sh
minting 10 to 9x
{
 "boostIds": [0]
}
```

File: test/src/helpers.ts [hunk 14]
```diff
@@ -852,6 +860,7 @@ export function fundErc20(
   return async function fundErc20() {
     if (!erc20) erc20 = await freshERC20();
     for (const address of [options.account.address, ...(funded ?? [])]) {
+      console.log(`minting ${amount} to ${address}`);
```
Comment (sammccord) [inline test/src/helpers.ts:863]:
```suggestion
```

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 3b3b4b16269c868f2e8bf8062d57863880c903ef
</p>

File: test/src/helpers.ts (hunk 14 above)
Comment (topocount) [reply test/src/helpers.ts:863]: I want to keep this in for now, at least during early dogfooding

File: packages/cli/src/commands/seed.ts (hunk 13 above)
Comment (topocount) [reply packages/cli/src/commands/seed.ts:426]: ah got it

File: packages/cli/src/commands/seed.ts (hunk 6 above)
Comment (topocount) [reply packages/cli/src/commands/seed.ts:132]: if a budget config is the same across templates, use the same one

Comment (topocount) [reply packages/cli/src/commands/seed.ts:132]: i suppose I should do deep equality

File: packages/cli/src/commands/seed.ts (hunk 7 above)
Comment (topocount) [reply packages/cli/src/commands/seed.ts:187]: I think a simple Boost Id is most immediately helpful for frontend devs looking for state but yes I agree.

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e914351eaf64d4fb2642503f3b16dc4f346c863f
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7b49e446e658a8f01ec95fdd5880a1102fa68124
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7b49e446e658a8f01ec95fdd5880a1102fa68124
</p>

File: packages/cli/src/commands/seed.ts [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+  }
+}
+
+async function fundBudget(
+  options: DeployableTestOptions,
+  erc20: MockERC20,
+  budget: Budget,
+  amount = parseEther('110'),
+) {
+  console.warn(`minting ${amount} to ${options.account.address}`);
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:420]:
```suggestion
```

File: packages/cli/src/commands/seed.ts [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+  criteriaType: z.nativeEnum(SignatureType),
+  signature: zAbiItemSchema,
+  fieldIndex: z.number().nonnegative(),
+  targetContract: AddressSchema,
+});
+
+export const ERC20VariableCriteriaIncentiveSchema = z.object({
+  type: z.literal('ERC20VariableCriteriaIncentive'),
+  asset: AddressSchema,
+  mintableMockAsset: z.boolean().default(true),
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:366]: i think this name would be clearer if it was something like `shouldMintAndAllocate` or `autoMintAndFund`

File: packages/cli/src/commands/seed.ts [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+      denied: [],
+    },
+    incentives: [
+      {
+        type: 'ERC20Incentive',
+        asset: asset,
+        strategy: 0,
+        reward: 1n,
+        limit: 1n,
+        manager: account,
```
Comment (sammccord) [inline packages/cli/src/commands/seed.ts:494]:
```suggestion
        manager: account,
        shouldMintAndAllocate: false
```

just to make this API obvious when they generate a seed

File: test/src/helpers.ts [hunk 18]
```diff
@@ -852,6 +860,7 @@ export function fundErc20(
   return async function fundErc20() {
     if (!erc20) erc20 = await freshERC20();
     for (const address of [options.account.address, ...(funded ?? [])]) {
+      console.warn(`minting ${amount} to ${address}`);
```
Comment (sammccord) [inline test/src/helpers.ts:863]:
```suggestion
```

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2eaab117ddf81bd85d940eb13e48154e48e2fbec
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 628f7c405cd51f0331d3abb460cf0a04abffe583
</p>

**Review by sammccord — APPROVED** [review]

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  By using the `[ADHOC]` title prefix, you are bypassing best practice protections.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5c403de0136255e1277f1d8bec54a98516a0bebc
</p>

**Pull request merged**